base/utfconvert.jl

```diff
@@ -0,0 +1,546 @@
+    @inbounds return UTF16String(setindex!(copy!(Vector{UInt16}(len+1),1,data,1,len),0,len+1))
+end
+
+convert(T::Type{UTF16String}, data::AbstractArray{UInt16}) =
+    convert(T, reshape(data, length(data)))
+
+convert(T::Type{UTF16String}, data::AbstractArray{Int16}) =
+    convert(T, reinterpret(UInt16, data))
+
+function convert(T::Type{UTF16String}, bytes::AbstractArray{UInt8})
+    isempty(bytes) && return UTF16String(UInt16[0])
+    isodd(length(bytes)) && throw(ArgumentError("odd number of bytes"))
+    data = reinterpret(UInt16, bytes)
+    # check for byte-order mark (BOM):
+    if data[1] == 0xfeff        # native byte order
+        d = Array(UInt16, length(data))
+        copy!(d,1, data,2, length(data)-1)
+    elseif data[1] == 0xfffe    # byte-swapped
+        d = Array(UInt16, length(data))
+        for i = 2:length(data)
+            d[i-1] = bswap(data[i])
+        end
+    else
+        d = Array(UInt16, length(data) + 1)
+        copy!(d,1, data,1, length(data)) # assume native byte order
+    end
+    d[end] = 0 # NULL terminate
+    !isvalid(UTF16String, d) && throw(ArgumentError("invalid UTF16 data"))
+    UTF16String(d)
+end
+
+utf16(p::Ptr{UInt16}, len::Integer) = utf16(pointer_to_array(p, len))
+utf16(p::Ptr{Int16}, len::Integer) = utf16(convert(Ptr{UInt16}, p), len)
+function utf16(p::Union(Ptr{UInt16}, Ptr{Int16}))
+    len = 0
+    while unsafe_load(p, len+1) != 0; len += 1; end
+    utf16(p, len)
+end
+
+function convert(::Type{UTF32String}, data::AbstractVector{Char})
+    len = length(data)
+    @inbounds return UTF32String(setindex!(copy!(Vector{Char}(len+1),1,data,1,len),0,len+1))
+end
+
+convert{T<:Union(Int32,UInt32)}(::Type{UTF32String}, data::AbstractVector{T}) =
+    convert(UTF32String, reinterpret(Char, data))
+
+convert{T<:AbstractString}(::Type{T}, v::AbstractVector{Char}) = convert(T, utf32(v))
+
+# specialize for performance reasons:
+function convert{T<:ByteString}(::Type{T}, data::AbstractVector{Char})
+    s = IOBuffer(Array(UInt8,length(data)), true, true)
+    truncate(s,0)
+    for x in data
+        print(s, x)
+    end
+    convert(T, takebuf_string(s))
+end
+
+convert(::Type{Vector{Char}}, str::UTF32String) = str.data
+convert(::Type{Array{Char}},  str::UTF32String) = str.data
+
+unsafe_convert{T<:Union(Int32,UInt32,Char)}(::Type{Ptr{T}}, s::UTF32String) =
+    convert(Ptr{T}, pointer(s))
+
+function convert(T::Type{UTF32String}, bytes::AbstractArray{UInt8})
+    isempty(bytes) && return UTF32String(Char[0])
+    length(bytes) & 3 != 0 && throw(ArgumentError("need multiple of 4 bytes"))
+    data = reinterpret(Char, bytes)
+    # check for byte-order mark (BOM):
+    if data[1] == Char(0x0000feff) # native byte order
+        d = Array(Char, length(data))
+        copy!(d,1, data, 2, length(data)-1)
+    elseif data[1] == Char(0xfffe0000) # byte-swapped
+        d = Array(Char, length(data))
+        @inbounds for i = 2:length(data) ; d[i-1] = bswap(data[i]) ; end
+    else
```

Comment:
No, it isn't generally, and also see my issues #11501 and #11502, about what I'd like to see done here.